src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ValueProviderResult.cs

```diff
@@ -142,9 +142,17 @@ private object ConvertSimpleType(CultureInfo culture, object value, Type destina
             if (!(canConvertFrom || converter.CanConvertTo(destinationType)))
             {
                 // EnumConverter cannot convert integer, so we verify manually
-                if (destinationType.GetTypeInfo().IsEnum && (value is int))
+                if (destinationType.GetTypeInfo().IsEnum &&
+                    (value is int ||
+                    value is uint ||
+                    value is long ||
+                    value is ulong ||
+                    value is short ||
+                    value is ushort ||
+                    value is byte ||
+                    value is sbyte))
                 {
-                    return Enum.ToObject(destinationType, (int)value);
+                    return Enum.ToObject(destinationType, Convert.ToInt32(value));
```

Comment:
> I trust that HtmlHelpers do the right thing with non-defined values?

Yes they do the right thing. They ignore whether or not the value is defined. Same as it ever was because a defined `enum` value is just an `int` (or similar) with a special name.